### PR TITLE
Implement migrations for RayyanStore reviews

### PR DESCRIPTION
**Pull request by fabubaker**
- Add script (scaffold-version.js) to generate migration skeleton.
- Use migrator to create and upgrade reviews.
- Setup tests using Jest.
- Add endpoint to get the version of a review.
- Modify createReview endpoint to accept version as a parameter.
- Update Ruby client to reflect API changes, bump version to 0.1.16

### CODE REVIEW

File: server/src/app.ts [hunk 1]
```diff
@@ -40,6 +41,32 @@ app.use(function (req, res, next) {
   next()
 })
 
+app.use(function (req, res, next) {
+  // Exclude createReview and deleteReview from auto-migrations
+  const re = new RegExp('/reviews/[0-9]*');
+
+  if (re.test(req.path)) {
```
Comment (hammady) [inline server/src/app.ts:48]: We may support a `GET` request that matches this path in the future, which is neither `createReview` or `deleteReview`. I suggest explicitly checking for `POST` and `DELETE` methods to make it future-proof.

File: server/src/app.ts [hunk 2]
```diff
@@ -40,6 +41,32 @@ app.use(function (req, res, next) {
   next()
 })
 
+app.use(function (req, res, next) {
```
Comment (hammady) [inline server/src/app.ts:44]: Does this middleware deserve to exist in its own file?

File: server/src/app.ts [hunk 3]
```diff
@@ -40,6 +41,32 @@ app.use(function (req, res, next) {
...
+  // Exclude createReview and deleteReview from auto-migrations
+  const re = new RegExp('/reviews/[0-9]*');
+
+  if (re.test(req.path)) {
+    return next()
+  }
+
+  // Extract review ID from request
+  let reviewId = req.query.review_id || req.body.review_id || _.last(req.url.split('/'))
```
Comment (hammady) [inline server/src/app.ts:53]: I thought we now attach `reviewId` as an HTTP X-Header?
This will also ignore `createReview`?

File: server/src/app.ts [hunk 4]
```diff
@@ -40,6 +41,32 @@ app.use(function (req, res, next) {
...
+  reviewId = parseInt(reviewId)
+
+  if (reviewId) {
+    adapter.handleMigrations(reviewId)
+    .then(function() {
+      next()
+    })
+    .catch(function (err) {
+      console.log(err)
+      res.status(500).json({ error: err.message })
```
Comment (hammady) [inline server/src/app.ts:63]: I am trying to think, is it safe to ignore the migration error and move on to the request as a best-effort approach? In some (most?) cases, the request will succeed, and if it fails, it was destined to fail anyway.

File: server/src/adapters/postgres/migrations/20200906112019_create_initial_rayyan_store_tables.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+import { Pool, PoolClient } from 'pg'
+import { BaseMigration } from '../migrator'
+import { fulltextKeys, singleValuedKeys, multiValuedKeys, sortKeys } from '../shared'
+
+export class CreateInitialRayyanStoreTables implements BaseMigration {
+  static version = 20200906112019
+
+  static async upgrade(reviewId: number, client: Pool|PoolClient): Promise<any> {
+    // Fill in upgrade commands here
+    const queryText = `
```
Comment (hammady) [inline server/src/adapters/postgres/migrations/20200906112019_create_initial_rayyan_store_tables.ts:11]: I assumed this is a blind copy from the old file.

File: server/src/adapters/postgres/migrator/base-migration.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * Maps a version number to a migration.
+ */
+export interface versionMapping {
+  [version: number]: typeof BaseMigration
+}
+
+/**
+ * Base class for implementing migrations.
+ */
+export class BaseMigration {
```
Comment (hammady) [inline server/src/adapters/postgres/migrator/base-migration.ts:13]: Do we need a `downgrade` function?

File: server/src/adapters/postgres/postgres.ts [hunk 7]
```diff
@@ -541,260 +519,70 @@ module.exports = {
...
-        DROP TABLE IF EXISTS r${reviewId}_customizations_log;
-        DROP SEQUENCE IF EXISTS r${reviewId}_customizations_log_id_seq;
-        DROP TABLE IF EXISTS r${reviewId};
-      `).then(function (res) {
+    // Remove version and delete review in a transaction.
+    const client = await pool.connect()
+    try {
+      await client.query('BEGIN')
+
+      migrator.removeVersion(reviewId, client)
```
Comment (hammady) [inline server/src/adapters/postgres/postgres.ts:549]: Do we also remove the version when soft deleting?

**Review by hammady — CHANGES_REQUESTED** [review]

File: server/src/adapters/postgres/migrations/index.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import { versionMapping } from '../migrator'
+
+export const versionMappings: versionMapping = {}
+
+import { CreateInitialRayyanStoreTables } from './20200906112019_create_initial_rayyan_store_tables'
```
[the remark this answers is not in the document]
Comment (fabubaker) [reply server/src/adapters/postgres/migrations/index.ts:5]: I adapted the Rails schema of naming migrations to the date that they were generated using the `scaffold-version` script. The format is `YYYYMMDDHHmmss`. 

Do you want me to change it to a linear versioning scheme?

File: server/src/app.ts (hunk 2 above)
Comment (fabubaker) [reply server/src/app.ts:44]: Isn't moving it to its own file a bit much? What's the reasoning?

File: server/src/app.ts (hunk 3 above)
Comment (fabubaker) [reply server/src/app.ts:53]: `_.last(req.url.split('/')` takes care of createReview, but isn't `X-Review-ID` a rayyan-store-proxy detail? Although I do think it's easier to use than attempting to manually retrieve the reviewId here. 

I'll switch to `X-Review-ID`.

File: server/src/adapters/postgres/migrations/20200906112019_create_initial_rayyan_store_tables.ts (hunk 5 above)
Comment (fabubaker) [reply server/src/adapters/postgres/migrations/20200906112019_create_initial_rayyan_store_tables.ts:11]: Yep.

File: server/src/adapters/postgres/postgres.ts (hunk 7 above)
Comment (fabubaker) [reply server/src/adapters/postgres/postgres.ts:549]: Hmm, you're saying we should "soft-delete" the version as well?

File: server/src/app.ts (hunk 4 above)
Comment (fabubaker) [reply server/src/app.ts:63]: My worry with this is that it might lead to some data-integrity issues. I think it's best for migrations to fail fast and hard.

File: server/src/adapters/postgres/migrations/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (hammady) [reply server/src/adapters/postgres/migrations/index.ts:5]: The rails schema is `date_random` to automate the sequence (the `date` part) and to guarantee uniqueness if 2 processes/developers created 2 migrations at the same second. If we are to follow this schema we should be aware that these are problems we need to fix. Yes, the `date` is the easiest way to automate the sequence and it will be cumbersome to generate unique sequences otherwise. Is there a way to have all zeros and not break this schema? Do you verify it somewhere?

File: server/src/app.ts (hunk 2 above)
Comment (hammady) [reply server/src/app.ts:44]: Typically middlewares are best maintained in their own files, but this one might be too small to qualify. I'm only drawing your attention to this possibility. Don't do if you see it an overkill.

File: server/src/app.ts (hunk 3 above)
Comment (hammady) [reply server/src/app.ts:53]: No it is not an RSP detail, it is a client specification as exposed by the gem. RSP only consumes it and RS can do this too.

File: server/src/app.ts (hunk 4 above)
Comment (hammady) [reply server/src/app.ts:63]: That might be true. But I cannot overlook the situation where we would have a bug in migrations that will fail the majority of requests where they would normally succeed without attempting migrations.

File: server/src/adapters/postgres/migrator/base-migration.ts (hunk 6 above)
Comment (hammady) [reply server/src/adapters/postgres/migrator/base-migration.ts:13]: Before rushing into the implementation, do we need it now? How would we invoke it?

File: server/src/adapters/postgres/postgres.ts (hunk 7 above)
Comment (hammady) [reply server/src/adapters/postgres/postgres.ts:549]: Yes, because not doing so will leave the soft-deleted review in an incomplete state (you don't know what version it corresponds to). So either soft-delete the version as well (add a boolean: deleted), or don't delete it at all, what is the impact of leaving it? It would be hard to clean it I guess.

File: server/src/adapters/postgres/migrator/base-migration.ts (hunk 6 above)
Comment (fabubaker) [reply server/src/adapters/postgres/migrator/base-migration.ts:13]: We don't need it now but it might be better to have downgrade functions ready in case we do need them in the future. Although I can't think of a scenario where we would need them...

File: server/src/adapters/postgres/migrations/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (fabubaker) [reply server/src/adapters/postgres/migrations/index.ts:5]: I guess technically you could have the base version be all zeroes and it would still work with the migrator as it is. See the table definition, for eg:

```
CREATE TABLE IF NOT EXISTS review_versions (
        review_id INTEGER PRIMARY KEY,
        version BIGINT CHECK (version >= 0)
      )
```

I can change it to `0000000000000_create_initial_rs_tables`, but we are sticking to the date as the migration sequence yes?

File: server/src/adapters/postgres/scripts/scaffold-version.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env node
```
[the remark this answers is not in the document]
Comment (fabubaker) [reply server/src/adapters/postgres/scripts/scaffold-version.js:1]: Done, I added a `server` section to the readme.

File: server/src/app.ts (hunk 1 above)
Comment (fabubaker) [reply server/src/app.ts]: Done.

File: server/src/app.ts (hunk 4 above)
Comment (fabubaker) [reply server/src/app.ts:62]: Isn't that what tests and staging is for? :)

File: server/src/adapters/postgres/postgres.ts (hunk 7 above)
Comment (fabubaker) [reply server/src/adapters/postgres/postgres.ts:549]: The problem is that there can be multiple soft-deleted versions of the same review, so we may need a separate table for them.

Is soft-deleting even useful anymore? I'm in favor of getting rid of it.

File: server/src/adapters/postgres/migrations/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (hammady) [reply server/src/adapters/postgres/migrations/index.ts]: Yes. Please go ahead.

File: server/src/adapters/postgres/scripts/scaffold-version.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (hammady) [reply server/src/adapters/postgres/scripts/scaffold-version.js:1]: In the new section, you specified `npm` as the migration command and `docker-compose` as the test runner. Either use compose in all or none.

File: server/src/app.ts (hunk 4 above)
Comment (hammady) [reply server/src/app.ts:62]: Yes, but only if we guarantee bugs do not show up later in some edge cases!

File: server/src/adapters/postgres/migrator/base-migration.ts (hunk 6 above)
Comment (hammady) [reply server/src/adapters/postgres/migrator/base-migration.ts:13]: Isn't deleting a review a downgrade of version 0? Can you implement it as such?

File: server/src/adapters/postgres/postgres.ts (hunk 7 above)
Comment (hammady) [reply server/src/adapters/postgres/postgres.ts:549]: OK, get rid of it. Later we can automatically archive reviews before deletion (could be out of scope of rayyan-store, but its clients).

File: server/src/adapters/postgres/migrations/index.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import { versionMapping } from '../migrator'
+
+export const versionMappings: versionMapping = {}
+
+import { CreateInitialRayyanStoreTables } from './00000000000000_create_initial_rayyan_store_tables'
```
Comment (hammady) [inline server/src/adapters/postgres/migrations/index.ts:5]: This looks very neat to my eyes now!

File: server/src/adapters/postgres/migrator/base-migration.ts (hunk 6 above)
Comment (fabubaker) [reply server/src/adapters/postgres/migrator/base-migration.ts:13]: I thought about initially implementing `deleteReview` as a series of cascaded downgrades, but it's just simpler to drop the tables.

I propose not having a downgrade for version 0 (maybe have it throw an error), and then I can have downgrades for all the other versions, if we ever happen to need them.

File: server/src/adapters/postgres/scripts/scaffold-version.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (fabubaker) [reply server/src/adapters/postgres/scripts/scaffold-version.js:1]: Done.

File: server/src/adapters/postgres/postgres.ts (hunk 7 above)
Comment (fabubaker) [reply server/src/adapters/postgres/postgres.ts:547]: Removed.

File: README.md [hunk 11]
```diff
@@ -59,7 +59,7 @@ To generate a migration, run the following:
 
 ```
 cd server
-npm run migrator:scaffold "<title of migration>"
+docker-compose run --rm -v $PWD:/home server npm run migrator:scaffold "<title of migration>"
```
Comment (hammady) [inline README.md:62]: Why not declare the volume inside the `docker-compose.yaml`? We are getting rid of compose soon anyway, so this is optional if it takes more than a minute.

Comment (fabubaker) [reply README.md]: Done

File: server/src/app.ts (hunk 4 above)
Comment (hammady) [reply server/src/app.ts:62]: Agreed to fail early.

File: server/src/adapters/postgres/migrator/base-migration.ts (hunk 6 above)
Comment (hammady) [reply server/src/adapters/postgres/migrator/base-migration.ts:13]: Agreed that it is simpler to delete reviews explicitly instead of undoing all migrations one by one.

**Review by hammady — APPROVED** [review]

**Review by hammady — COMMENTED** [review]
Any endpoint to invoke a downgrade?

**Comment by fabubaker** [comment]
> Any endpoint to invoke a downgrade?

Yep, done.

File: server/src/adapters/postgres/migrator/migrator.ts [hunk 12]
```diff
@@ -77,6 +77,43 @@ export class Migrator {
     }
   }
 
+  async downgrade(reviewId: number, downgradeVersion: number): Promise<any> {
```
Comment (hammady) [inline server/src/adapters/postgres/migrator/migrator.ts:80]: What happens if there is another request targeting the same review which will try to upgrade in the same time? If we lock the tables, the upgrade will be sequenced, then what is the benefit of downgrade?

File: server/src/app.ts [hunk 13]
```diff
@@ -115,6 +115,21 @@ app.get('/reviews/:review_id/version', (req, res) => {
     .then(function (version) {
       res.json(version)
     })
+    .catch(function (err) {
+      console.log(err)
+      res.status(500).json({ status: err.message })
+    })
+})
+
+app.post('/reviews/:review_id/downgrade', (req, res) => {
```
Comment (hammady) [inline server/src/app.ts:124]: Will this request skip the upgrade middleware?

Comment (fabubaker) [reply server/src/app.ts:124]: Good catch, I should also disable it for the `getReviewVersion` endpoint.

File: server/src/adapters/postgres/migrator/migrator.ts (hunk 12 above)
Comment (fabubaker) [reply server/src/adapters/postgres/migrator/migrator.ts:80]: It will be useful in scenarios where the latest migration has to be withdrawn and all reviews have to be dowgraded to an older version.

Comment (hammady) [reply server/src/adapters/postgres/migrator/migrator.ts:83]: In which case, we have to change the latest version back to prevent further upgrades, then we launch downgrades. Does this mean the race condition I mentioned above cannot happen?

File: server/src/app.ts [hunk 14]
```diff
@@ -43,7 +43,7 @@ app.use(function (req, res, next) {
 
 app.use(function (req, res, next) {
   // Exclude createReview and deleteReview from auto-migrations
```
Comment (hammady) [inline server/src/app.ts:45]: If this covers downgrades and getVersion, then let's document it here.

File: server/src/adapters/postgres/migrator/migrator.ts (hunk 12 above)
Comment (fabubaker) [reply server/src/adapters/postgres/migrator/migrator.ts:83]: Assuming we change the latest version back, any attempt to perform an automatic upgrade will fail (because `upgradeVersion < currentVersion`). So only the downgrade will go through.

Comment (hammady) [reply server/src/adapters/postgres/migrator/migrator.ts:83]: Why does it imply removing the migration itself? In case of rollback, we just change the latest version and keep the migrations so that downgrades work. Would that work?

Comment (fabubaker) [reply server/src/adapters/postgres/migrator/migrator.ts:83]: The latest version is automatically obtained from the sorted list of versions:
```
this.latestVersion = this.versions[this.versions.length - 1]
```

So the only way to change the latest version is to not export it from the `migrations` folder.

Comment (hammady) [reply server/src/adapters/postgres/migrator/migrator.ts:83]: I thought we explicitly set a variable for the last version. This gives more flexibility but any new migration PR should also update this variable (or the generator script).

Comment (fabubaker) [reply server/src/adapters/postgres/migrator/migrator.ts:83]: How about something like:

```
this.latestVersion = parseInt(process.env.LATEST_MIGRATION_VERSION || this.versions[this.versions.length - 1])
```

So if it's not specified, the latest version defaults to the last version in the list.

Comment (fabubaker) [reply server/src/adapters/postgres/migrator/migrator.ts:83]: Helm overrides should be good no?

Comment (hammady) [reply server/src/adapters/postgres/migrator/migrator.ts:86]: I thought you will always set the latest version and every migration class generator should also change it. This is as asking the developer go check the migrations folder to determine the latest instead of looking at a number in the same file. Anyway, it servers the purpose.

**Review by hammady — APPROVED** [review]
